src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/PooledTxsRequestor.cs

```diff
@@ -15,17 +15,33 @@
 
 namespace Nethermind.Network.P2P.Subprotocols.Eth
 {
-    public class PooledTxsRequestor(ITxPool txPool, ITxPoolConfig txPoolConfig, ISpecProvider specProvider) : IPooledTxsRequestor
+    public class PooledTxsRequestor : IPooledTxsRequestor, IDisposable
     {
         private const int MaxNumberOfTxsInOneMsg = 256;
-        private readonly bool _blobSupportEnabled = txPoolConfig.BlobsSupport.IsEnabled();
-        private readonly long _configuredMaxTxSize = txPoolConfig.MaxTxSize ?? long.MaxValue;
+        private readonly bool _blobSupportEnabled;
+        private readonly long _configuredMaxTxSize;
 
-        private readonly long _configuredMaxBlobTxSize = txPoolConfig.MaxBlobTxSize is null
+        private readonly long _configuredMaxBlobTxSize;
+
+        private readonly ClockKeyCache<ValueHash256> _pendingHashes = new(MemoryAllowance.RequestedTxHashCacheSize);
+        private readonly ITxPool _txPool;
+
+        public PooledTxsRequestor(ITxPool txPool, ITxPoolConfig txPoolConfig, ISpecProvider specProvider)
+        {
+            _txPool = txPool;
+            _blobSupportEnabled = txPoolConfig.BlobsSupport.IsEnabled();
+            _configuredMaxTxSize = txPoolConfig.MaxTxSize ?? long.MaxValue;
+            _configuredMaxBlobTxSize = txPoolConfig.MaxBlobTxSize is null
             ? long.MaxValue
             : txPoolConfig.MaxBlobTxSize.Value + (long)specProvider.GetFinalMaxBlobGasPerBlock();
 
-        private readonly ClockKeyCache<ValueHash256> _pendingHashes = new(MemoryAllowance.TxHashCacheSize);
+            _txPool.TxPoolHeadChanged += TxPoolHeadChanged;
+        }
+
+        private void TxPoolHeadChanged(object? sender, Block e)
+        {
+            _pendingHashes.Clear();
```

Comment:
Can't this cause that we don't request some transaction hashes?

Reply:
Pool kicks out some txs and then marks them not long term blocked, but blocked in this block; so we want to re-request them. Currently this second filter in the requestor will block that

Also these are marked when we request a tx; so as not to repeat requests; however we don't know if we didn't get a response for the request; currently they are blocked just by being requested; so now each head we clear that, so will rerequest ones that never got a response.

One of the filters is if we have in pool (or txpool has marked not interesting); so won't be requesting everything.

> we want to re-request transactions that are probably less valuable than the ones in pool?

Have a gap in pool of 100+ txs after head; so would want the ones we had previously rejected to fill

Reply:
Ok can you show metrics of tx request messages before and after?